File: app/src/main/java/com/example/myapplication11/ui/home/HomeFragment.java
package com.example.myapplication11.ui.home;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.myapplication11.R;
import com.example.myapplication11.base.BaseFragment;
import com.example.myapplication11.bean.responsebean.ArticleBean;
import com.example.myapplication11.bean.responsebean.home.HomeData;
import com.example.myapplication11.databinding.FragmentListBinding;
import com.example.myapplication11.navinterface.ScrollToTop;
import com.example.myapplication11.ui.adapter.HomeAdapter;
import com.scwang.smart.refresh.layout.api.RefreshLayout;
import com.scwang.smart.refresh.layout.listener.OnLoadMoreListener;
import com.scwang.smart.refresh.layout.listener.OnRefreshListener;
import com.zhouwei.mzbanner.MZBannerView;

import java.util.List;


public class HomeFragment extends BaseFragment<FragmentListBinding, HomeViewModel> implements ScrollToTop {


    private HomeAdapter commonAdapter;

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_list;
    }

    @Override
    protected void initViewModel() {
        mViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
        initDataChange();


    }

    @Override
    protected void bindViewModel() {

    }

    @Override
    protected boolean isSupportLoad() {
        return true;
    }

    @Override
    protected void init() {
        mViewModel.loadHomeData();

        initRecycle();
        initRefreshLayout();

    }


    private void initRefreshLayout() {

        mDataBinding.refreshLayout.setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                mViewModel.refreshData(true);
            }
        });


        mDataBinding.refreshLayout.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                mViewModel.refreshData(false);
            }
        });
    }


    private void initRecycle() {
        commonAdapter = new HomeAdapter() {
            @Override
            public void addListener(View root, HomeData itemData, int position) {
                super.addListener(root, itemData, position);
                if (itemData.getBannerData() != null) {
                    MZBannerView banner = root.findViewById(R.id.banner);
                    banner.setBannerPageClickListener(new MZBannerView.BannerPageClickListener() {
                        @Override
                        public void onPageClick(View view, int i) {
                            String url = itemData.getBannerData().getBannerData().get(i).getUrl();
                        }
                    });
                } else if (itemData.getTopArticleList() != null) {

                } else {
                    root.findViewById(R.id.card_view).setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {

                        }
                    });

                    root.findViewById(R.id.iv_collect).setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            itemData.getArticleList().setCollect(!itemData.getArticleList().isCollect());
                            notifyDataSetChanged();
                            mViewModel.changeArticleCollect(itemData.getArticleList().isCollect(), itemData.getArticleList().getId());
                        }
                    });
                }
            }


            @Override
            public void addTopClickListener(ArticleBean itemData) {
                super.addTopClickListener(itemData);

            }

            @Override
            public void addTopCollectListener(ArticleBean itemData) {
                super.addTopCollectListener(itemData);
                mViewModel.changeArticleCollect(itemData.isCollect(), itemData.getId());
            }
        };
        mDataBinding.recycle.setLayoutManager(new LinearLayoutManager(getContext()));
        mDataBinding.recycle.setAdapter(commonAdapter);
    }

    private void initDataChange() {
        mViewModel.getHomeList().observe(this, new Observer<List<HomeData>>() {
            @Override
            public void onChanged(List<HomeData> homeData) {
                commonAdapter.onItemDatasChanged(homeData);
                mDataBinding.refreshLayout.finishRefresh();
                mDataBinding.refreshLayout.finishLoadMore();
            }
        });
    }

    @Override
    public void scrollToTop() {
        mDataBinding.recycle.smoothScrollToPosition(0);
    }
}

File: appkotlin/src/main/java/com/example/appkotlin/activity/ui/home/HomeFragment.kt
package com.example.appkotlin.activity.ui.home

import android.content.Context
import android.graphics.Color
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.bumptech.glide.Glide
import com.example.appkotlin.MyApplication
import com.example.appkotlin.R
import com.example.appkotlin.base.BaseFragment
import com.example.appkotlin.bean.Data
import com.example.appkotlin.databinding.FragmentHomeBinding
import com.zhouwei.mzbanner.holder.MZHolderCreator
import com.zhouwei.mzbanner.holder.MZViewHolder
import kotlinx.android.synthetic.main.fragment_home.*


class HomeFragment : BaseFragment<FragmentHomeBinding, HomeViewModel>() {


    override fun getResLayoutId(): Int {
        return R.layout.fragment_home
    }

    override fun initViewModel() {
        mViewModel = ViewModelProvider(this).get(HomeViewModel::class.java)

    }

    override fun bindViewModel() {
//        mDataBinding.viewModel = mViewModel
//        mViewModel.loadState.postValue(LoadState.LOADING)

        mViewModel.loadData()

        mViewModel.mutableBanner.observe(this, Observer { bannerBean ->
            mzBanner.setPages(bannerBean.data as List<Nothing>?, MZHolderCreator<BannerViewHolder> { return@MZHolderCreator BannerViewHolder() })
        })


    }

    override fun init() {
    }


    class BannerViewHolder() : MZViewHolder<Data> {

        lateinit var imageView: ImageView

        override fun createView(p0: Context?): View {
            var view = LayoutInflater.from(p0).inflate(R.layout.item_banner, null)
            imageView = view.findViewById(R.id.img)
            return view
        }

        override fun onBind(context: Context?, p1: Int, data: Data?) {
            if (context != null) {
                Log.e("1--", "0" + (data?.imagePath));
                Glide.with(context).load(data?.imagePath).into(imageView)
            } else {
                Log.e("1--", "1");
            }
        }

    }


}


